### PR TITLE
Add overlay flag to IgnoreChownErrors

### PR DESCRIPTION
**Pull request by rhatdan**
Modified patch of Kevin Pelzel.

Also changed ApplyDiff to take new ApplyDiffOpts Struct.

Signed-off-by: Kevin Pelzel <kevinpelzel22@gmail.com>
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Goal is to Replace https://github.com/containers/storage/pull/358

**Comment by rhatdan** [comment]
@nalind @giuseppe @vrothberg PTAL

**Comment by rhatdan** [comment]
@kpelzel  PTAL

**Comment by hypery2k** [comment]
@rhatdan Can you somehow share the binaries? Trying to build it locally but didn't get it working:
```
Makefile:53: recipe for target 'images_ffjson.go' failed
make: *** [images_ffjson.go] Error 1
```

**Comment by rhatdan** [comment]
@hypery2k are you building in go/src/github.com/containers/storage?

**Comment by hypery2k** [comment]
yes, here's my snippet from my Dockerfile:
```
RUN export GOROOT=/usr/local/go && export PATH=$GOROOT/bin:$PATH && go get github.com/pquerna/ffjson && \
        mkdir -p /var/tmp/go/src/github.com/containers/ && cd /var/tmp/go/src/github.com/containers/ && git clone https://github.com/rhatdan/storage.git && cd storage && git checkout applydiff && \
        source contrib/cirrus/lib.sh && install_ooe && make install.tools && make vendor && showrun make local-binary
```

**Comment by hypery2k** [comment]
@rhatdan for a quick fix it would help my if you can share just the binary with me, via dropbox or similiar share...

Thanks in advance

**Comment by rhatdan** [comment]
I have no idea why this will not build for you.  Could you try to just checkout the images_ffjson.go file?

I don't have an easy way to build this for Ubuntu right now, and I am no PTO.
@vrothberg @nalind Could you help out @hypery2k ?

**Comment by vrothberg** [comment]
@hypery2k, maybe the generator is confused by local files. Can you try a `git clean -xdf` and then try again? It will remove all files not under version control.

**Comment by hypery2k** [comment]
doesn't help. Tried it in my Dockerfile: https://gist.github.com/hypery2k/0a1d1e9b39ce0cc84e87ce1150eafade

Any way to share the patched binaries for Centos/RHEL?

**Comment by vrothberg** [comment]
@hypery2k, you can use a branch in my libpod fork: https://github.com/vrothberg/libpod/tree/applydiff

No need to change anything, it has already vendored this PR. Just `git clone` from there, and `make podman`.

**Comment by hypery2k** [comment]
thanks that worked for me

File: storage.conf [hunk 1]
```diff
@@ -32,6 +32,15 @@ size = ""
 # mountopt specifies comma separated list of extra mount options
 mountopt = "nodev"
 
+# ignore_chown_errors if using a user namespace without only a one UID, he can
+# setup storage with this flag to true. Then the user can pull and use any
+# image even those with multiple uids.  Note multiple UIDs will be squashed
+# down to the default uid in the container.  These images will have no
+# separation between the users in the container. Only supported for the overlay
+# driver.
+**ignore_chown_errors** = "true|False"
```
Comment (nalind) [inline storage.conf]: This looks like markdown, but the configuration file doesn't use markdown.

**Comment by nalind** [comment]
This is missing the `layers_ffjson.go` updates to catch it up with changes being made to `layers.go`.

**Comment by kpelzel** [comment]
I can only get this to work by adding `--storage-opt "overlay2.ignore_chown_errors=true"` to my buildah command. Changing ignore-chown-errors in ~/.config/containers/storage.conf does nothing for me. Is this the intended functionality or is it just because there needs to be code changes in buildah?

**Comment by rhatdan** [comment]
Ok I found an issue.  Also added support for VFS.

**Comment by rhatdan** [comment]
@kpelzel Please try this out.  I have tested with podman and it seems to do the right thing now.

**Comment by kpelzel** [comment]
Great, seems to be working now

**Comment by hypery2k** [comment]
nice, any plans for getting it merged and released?

**Comment by rhatdan** [comment]
Yes I hope to get this merged soon.  I am attempting to get vfs support for this feature also, this is the hold up.  I need to get the tests to pass then I can get it merged.

The issue I am seeing when this is merged into podman is I get this weird error.

```
/tmp/podman pull fedora
Trying to pull docker.io/library/fedora...
Getting image source signatures
Copying blob 07b72cd62760 done
Copying config 2b74bf3d24 done
Writing manifest to image destination
Storing signatures
ERRO[0013] Error while applying layer: unable to decode ApplyLayer JSON response: invalid character 'D' looking for beginning of value 
  unable to decode ApplyLayer JSON response: invalid character 'D' looking for beginning of value
Trying to pull registry.fedoraproject.org/fedora...
Getting image source signatures
Copying blob e656b45cbffb done
Copying config f7a80bfc4a done
Writing manifest to image destination
Storing signatures
ERRO[0033] Error while applying layer: unable to decode ApplyLayer JSON response: invalid character 'D' looking for beginning of value 
  unable to decode ApplyLayer JSON response: invalid character 'D' looking for beginning of value
Trying to pull quay.io/fedora...
  error parsing HTTP 404 response body: invalid character '<' looking for beginning of value: "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 3.2 Final//EN\">\n<title>404 Not Found</title>\n<h1>Not Found</h1>\n<p>The requested URL was not found on the server.  If you entered the URL manually please check your spelling and try again.</p>\n"
```

**Comment by rhatdan** [comment]
Correctly vendoring this in fixes the issue I was having with Podman. Now just got to figure out what is breaking with ostree.

**Comment by rhatdan** [comment]
Tests are finally passing.  Merging.

**Pull request merged**